### DISCUSSION TITLE
Pulsar upgrade to 3.0.5 causes prometheus metrics timeouts on brokers

### DISCUSSION THREAD

**Discussion by justin-lathrop** [apache/pulsar#22897] [Q&A]
After performing an upgrade from pulsar 2.9.4 to pulsar 3.0.5 within a Kubernetes cluster using pulsar helm 3.0 chart the Prometheus Metrics stopped working via the pulsar-brokers.

The pulsar-broker logs show a constant stream of 500 responses with timeouts, and 302 redirects.

```
... INFO org.eclipse.jetty.server.RequestLog - ... "GET /metrics HTTP/1.1" 302 0 "-" "Prometheus/2.42.0" 0
... WARN org.apache.pulsar.broker.stats.prometheus.PulsarPrometheusMetricsServlet - Prometheus metrics request timed out
... INFO org.eclipse.jetty.server.RequestLog - ... "GET /metrics/ HTTP/1.1" 500 - "http:/1.2.3.4:8080/metrics" "Prometheus/2.42.0" 60001
```

Running `pulsar-admin broker-stats monitoring-metrics` does return metrics with values of data moving through.  But exec'd into the pulsar-broker-0 pod and `curl http://localhost:8080/metrics/" only times out.

The pulsar-proxy instance was also upgraded  as part of this and it is reporting metrics as expected with no timeouts by curling the pulsar-proxy-0 metrics endpoint like so.  `curl http://localhost:8080/metrics/` from within the pod.

The error reported seems to be at this line in the code, but at present I do not see any of the other possible logs, it just seems to timeout every time in the async process.  https://github.com/apache/pulsar/blob/branch-3.0/pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PulsarPrometheusMetricsServlet.java#L83

Any help would be greatly appreciated!

**Comment by justin-lathrop** [comment]
I deployed instead the 3.0.4 release and the broker metrics started working again.  I wonder if its related to some of the metrics related changes in this commit? Which was put into the 3.0.5 release.

https://github.com/apache/pulsar/commit/7009071b6d53bbc3d740ea99cdc0c010692679ab

**Reply by lhotari** [reply]
@justin-lathrop Yes, there are quite a few changes for metrics in 3.0.5 ([release notes](https://github.com/apache/pulsar/releases/tag/v3.0.5)). 

It would help the investigation to get a threaddump after the request for metrics is in progress. you can use `jstack -l 1` in the pod to get the thread dump.

**Reply by lhotari** [reply]
How large is the uncompressed size of your metrics?

You could try disabling gzip compression with `PULSAR_PREFIX_httpServerGzipCompressionExcludedPaths: "^.*$"` passed in the broker's configData. 
Would you be able to test that with 3.0.5 ?

**Reply by justin-lathrop** [reply]
Attached is the output of the jstack command run **while** the curl was waiting for metrics response.

[jstack.txt](https://github.com/user-attachments/files/15842001/jstack.txt)

Using pulsar 3.0.4 the pulsar-broker-0 metrics output size is 78KB with no compression.

The exclude gzip setting did not have an affect.  Screenshot showing it just to make sure I didn't miss something there.
![excluded-gzip](https://github.com/apache/pulsar/assets/978309/adf2243d-63bf-47f8-853d-a3e4b9e18415)

**Reply by lhotari** [reply]
The jstack didn't reveal any deadlocks. What is odd is the high number of threads. There seems to be 36 threads for each thread pool in many cases. This could cause some waste in Netty buffer pools and that might matter if your memory is limited. Thread locals will also waste a lot of memory with such high thread counts.

What k8s resource settings and what JVM args do you have? You could use `cat /proc/1/cmdline` in the pod to find out the command line.

What k8s version and flavour are you using? How about broker configuration?

**Reply by justin-lathrop** [reply]
`pulsar-broker-0` has `requests` but no `limits`:
```
    Requests:
      cpu:      200m
      memory:   512Mi
```

```
pulsar@pulsar-broker-0:/pulsar$ cat /proc/1/cmdline
/usr/bin/java-Dlog4j.shutdownHookEnabled=false-cp/pulsar/conf:::/pulsar/lib/*:-Dlog4j2.formatMsgNoLookups=true-Dlog4j.configurationFile=log4j2.yaml-Djute.maxbuffer=10485760-Djava.net.preferIPv4Stack=true-Dzookeeper.clientTcpKeepAlive=true-Dio.netty.tryReflectionSetAccessible=true--add-opensjava.base/java.io=ALL-UNNAMED--add-opensjava.base/java.util.zip=ALL-UNNAMED--add-opensjava.base/java.nio=ALL-UNNAMED--add-opensjava.base/jdk.internal.misc=ALL-UNNAMED--add-opensjava.base/sun.net=ALL-UNNAMED--add-opensjava.management/sun.management=ALL-UNNAMED--add-opensjdk.management/com.sun.management.internal=ALL-UNNAMED--add-opensjava.base/jdk.internal.platform=ALL-UNNAMED-Xms128m-Xmx256m-XX:MaxDirectMemorySize=256m-XX:+IgnoreUnrecognizedVMOptions-XX:+UseG1GC-XX:MaxGCPauseMillis=10-Dio.netty.leakDetectionLevel=disabled-Dio.netty.recycler.linkCapacity=1024-XX:+ParallelRefProcEnabled-XX:+UnlockExperimentalVMOptions-XX:+DoEscapeAnalysis-XX:ParallelGCThreads=4-XX:ConcGCThreads=4-XX:G1NewSizePercent=50-XX:+DisableExplicitGC-XX:-ResizePLAB-XX:+ExitOnOutOfMemoryError-XX:+PerfDisableSharedMem-Xlog:async-Xlog:gc*,safepoint:/pulsar/logs/pulsar_gc_%p.log:time,uptime,tags:filecount=10,filesize=20M-Dpulsar.allocator.exit_on_oom=true-Dio.netty.recycler.maxCapacityPerThread=4096-Dpulsar.log.appender=RoutingAppender-Dpulsar.log.dir=/pulsar/logs-Dpulsar.log.level=info-Dpulsar.log.root.level=info-Dpulsar.log.immediateFlush=false-Dpulsar.routing.appender.default=Console-Dlog4j2.is.webapp=false-Dpulsar.functions.process.container.log.dir=/pulsar/logs-Dpulsar.functions.java.instance.jar=/pulsar/instances/java-instance.jar-Dpulsar.functions.python.instance.file=/pulsar/instances/python-instance/python_instance_main.py-Dpulsar.functions.extra.dependencies.dir=/pulsar/instances/deps-Dpulsar.functions.instance.classpath=/pulsar/conf:::/pulsar/lib/*:-Dpulsar.functions.log.conf=/pulsar/conf/functions_log4j2.xml-Dbookkeeper.metadata.bookie.drivers=org.apache.pulsar.metadata.bookkeeper.PulsarMetadataBookieDriver-Dbookkeeper.metadata.client.drivers=org.apache.pulsar.metadata.bookkeeper.PulsarMetadataClientDriver-Dpulsar.log.file=pulsar-broker.logorg.apache.pulsar.PulsarBrokerStarter--broker-conf/pulsar/conf/broker.conf
```

The k8s flavour/version: `v1.26.1+rke2r1`

The broker configuration is the same when run either on the `3.0.5` or `3.0.4` releases.  Here is a snip of the `pulsar-broker` configmap:
```
  PF_containerFactory: k8s
  PF_functionRuntimeFactoryClassName: org.apache.pulsar.functions.runtime.kubernetes.KubernetesRuntimeFactory
  PF_functionRuntimeFactoryConfigs_expectedMetricsCollectionInterval: "30"
  PF_functionRuntimeFactoryConfigs_installUserCodeDependencies: "true"
  PF_functionRuntimeFactoryConfigs_pulsarRootDir: /pulsar
  PF_functionRuntimeFactoryConfigs_submittingInsidePod: "true"
  PF_kubernetesContainerFactory_expectedMetricsCollectionInterval: "30"
  PF_kubernetesContainerFactory_installUserCodeDependencies: "true"
  PF_kubernetesContainerFactory_jobNamespace: pulsar
  PF_kubernetesContainerFactory_pulsarRootDir: /pulsar
  PF_kubernetesContainerFactory_submittingInsidePod: "true"
  PF_maxPendingAsyncRequests: "250"
  PF_numFunctionPackageReplicas: "1"
  PF_pulsarFunctionsCluster: pulsar
  PF_runtimeCustomizerClassName: org.apache.pulsar.functions.runtime.kubernetes.BasicKubernetesManifestCustomizer
  PULSAR_GC: |
    -XX:+UseG1GC -XX:MaxGCPauseMillis=10 -Dio.netty.leakDetectionLevel=disabled -Dio.netty.recycler.linkCapacity=1024 -XX:+ParallelRefProcEnabled -XX:+UnlockExperimentalVMOptions -XX:+DoEscapeAnalysis -XX:ParallelGCThreads=4 -XX:ConcGCThreads=4 -XX:G1NewSizePercent=50 -XX:+DisableExplicitGC -XX:-ResizePLAB -XX:+ExitOnOutOfMemoryError -XX:+PerfDisableSharedMem
  PULSAR_MEM: -Xms128m -Xmx256m -XX:MaxDirectMemorySize=256m -XX:+IgnoreUnrecognizedVMOptions
  PULSAR_PREFIX_acknowledgmentAtBatchIndexLevelEnabled: "true"
  PULSAR_PREFIX_allowAutoTopicCreation: "false"
  PULSAR_PREFIX_allowAutoTopicCreationType: partitioned
  PULSAR_PREFIX_brokerDeleteInactivePartitionedTopicMetadataEnabled: "true"
  PULSAR_PREFIX_brokerDeleteInactiveTopicsEnabled: "false"
  PULSAR_PREFIX_brokerEntryMetadataInterceptors: org.apache.pulsar.common.intercept.AppendIndexMetadataInterceptor
  PULSAR_PREFIX_deduplicationEnabled: "false"
  PULSAR_PREFIX_enableReplicatedSubscriptions: "false"
  PULSAR_PREFIX_exposeTopicLevelMetricsInPrometheus: "true"
  PULSAR_PREFIX_forceDeleteNamespaceAllowed: "true"
  PULSAR_PREFIX_forceDeleteTenantAllowed: "true"
  PULSAR_PREFIX_httpServerGzipCompressionExcludedPaths: ^.*$
  PULSAR_PREFIX_kafkaNamespace: ingest
  PULSAR_PREFIX_kafkaTenant: platform
  PULSAR_PREFIX_kafkaTransactionCoordinatorEnabled: "true"
  PULSAR_PREFIX_managedLedgerDefaultAckQuorum: "2"
  PULSAR_PREFIX_managedLedgerDefaultEnsembleSize: "3"
  PULSAR_PREFIX_managedLedgerDefaultWriteQuorum: "3"
  PULSAR_PREFIX_messagingProtocols: kafka
  PULSAR_PREFIX_numHttpServerThreads: "12"
  PULSAR_PREFIX_protocolHandlerDirectory: ./protocols
  PULSAR_PREFIX_saslAllowedMechanisms: PLAIN
  PULSAR_PREFIX_systemTopicEnabled: "true"
  PULSAR_PREFIX_topicLevelPoliciesEnabled: "true"
  authenticationEnabled: "true"
  authenticationProviders: org.apache.pulsar.broker.authentication.AuthenticationProviderToken
  authorizationEnabled: "true"
  brokerClientAuthenticationPlugin: org.apache.pulsar.client.impl.auth.AuthenticationToken
  exposeTopicLevelMetricsInPrometheus: "true"
  functionsWorkerEnabled: "true"
  managedLedgerDefaultAckQuorum: "1"
  managedLedgerDefaultEnsembleSize: "1"
  managedLedgerDefaultWriteQuorum: "1"
  numHttpServerThreads: "8"
```

**Reply by lhotari** [reply]
@justin-lathrop Do you know whether [cgroup v1 or v2 is in use](https://unix.stackexchange.com/questions/471476/how-do-i-check-cgroup-v2-is-installed-on-my-machine)? What kernel version do you have? (`uname -a`) 

I'm wondering that CPU requests of 200m doesn't result in the JVM assuming that there's 1 core to use. The workaround is described here: https://github.com/apache/pulsar/issues/17815#issuecomment-1268547528 . You can pass `-XX:ActiveProcessorCount=1` to the JVM arguments (in `PULSAR_MEM`, for example).

**Reply by lhotari** [reply]
What does "java -XshowSettings:system -version" return in the container? According to [this resource](https://developers.redhat.com/articles/2022/04/19/java-17-whats-new-openjdks-container-awareness#), that should also show the detected cgroup version.

**Reply by justin-lathrop** [reply]
cgroup v1 and `grep cgroup /proc/filesystems` shows only `nodev   cgroup`.

```
pulsar@pulsar-broker-0:/pulsar$ uname -a
Linux pulsar-broker-0 3.10.0-1160.95.1.el7.x86_64 #1 SMP Mon Jul 24 13:59:37 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

```
pulsar@pulsar-broker-0:/pulsar$ java -XshowSettings:system -version
Operating System Metrics:
    Provider: cgroupv1
    Effective CPU Count: 36
    CPU Period: 100000us
    CPU Quota: -1
    CPU Shares: 204us
    List of Processors, 36 total:
    0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35
    List of Effective Processors, 36 total:
    0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35
    List of Memory Nodes, 1 total:
    0
    List of Available Memory Nodes, 1 total:
    0
    Memory Limit: Unlimited
    Memory Soft Limit: Unlimited
    Memory & Swap Limit: Unlimited
    Maximum Processes Limit: Unlimited

openjdk version "17.0.11" 2024-04-16
OpenJDK Runtime Environment Temurin-17.0.11+9 (build 17.0.11+9)
OpenJDK 64-Bit Server VM Temurin-17.0.11+9 (build 17.0.11+9, mixed mode, sharing)
```

Updating the `PULSAR_MEM` to have additionally `-XX:ActiveProcessorCount=1` then restarting the pulsar-broker sts continues to give the same metrics timeout.

**Reply by lhotari** [reply]
@justin-lathrop Do you have a chance to increase max heap size and max direct memory size to see if it makes the problem go away? Please test with `-Xmx512m -XX:MaxDirectMemorySize=512m`.

Side comments:
That is an ancient kernel version. Linux 3.10 kernel was released in 2013, obviously there are backports in the RHEL 7 kernel, but for running Kubernetes and Java, I'd recommend switching to a modern OS.

**Reply by lhotari** [reply]
How many topics do you have? If you change `PULSAR_PREFIX_exposeTopicLevelMetricsInPrometheus: "true"` to `PULSAR_PREFIX_exposeTopicLevelMetricsInPrometheus: "false"`, does it make a difference?

**Reply by justin-lathrop** [reply]
It is `ancient` I agree, but its my dev cluster and hasn't been rebuilt in quite a while.  This cluster is mostly unused, it has `12` total topics all of which have no data going into them.

On a different cluster where this was found the `PULSAR_PREFIX_exposeTopicLevelMetricsInPrometheus: "false"` config was tried  prior to this initial post, and it did not fix the problem.  I check on that cluster's info out of curiosity and its RedHat 8, with still cgroups v1.

**Reply by justin-lathrop** [reply]
I tried increasing the `PULSAR_MEM` to `-Xmx512m -XX:MaxDirectMemorySize=512m` also with `PULSAR_PREFIX_exposeTopicLevelMetricsInPrometheus: "false"` then rollout restart the pulsar-broker sts, the metrics still hang though.

**Reply by lhotari** [reply]
Since it's a dev cluster, perhaps you could enable debug logging for Pulsar brokers to see if it would reveal some useful details about the problem with metrics. You can enable debug logging by setting `PULSAR_LOG_LEVEL: debug` in the broker's environment (in broker.configData).

**Reply by lhotari** [reply]
>  I check on that cluster's info out of curiosity and its RedHat 8, with still cgroups v1.

yes, RHEL 8 defaults to cgroups v1, but [it's possible to switch to cgroups v2](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/managing_monitoring_and_updating_the_kernel/index#mounting-cgroups-v2_using-cgroups-v2-to-control-distribution-of-cpu-time-for-applications). I don't think that there's much advantage of switching. What seems to matter more with Kubernetes and Java is having a fairly recent kernel version. 
Btw. It's recommended to configured Kubernetes notes with THP setting set to madvise. This is what you get also on Cloud provider managed Kubernetes nodes. Azul has a good guide [how to configure THP this for RHEL](https://docs.azul.com/prime/Enable-Huge-Pages#permanent-thp-setting-on-rhel-centos-oracle-linux-amazon-linux) and others. It's recommended to configure THP this way also when huge pages aren't used. The default `/sys/kernel/mm/transparent_hugepage/enabled` setting of `always` is bad for running Java unless `-XX:+AlwaysPreTouch` and `MALLOC_ARENA_MAX=2` is used.

**Reply by justin-lathrop** [reply]
Hey @lhotari sorry its been a bit.  But I was able to circle back to this and realized I am using [kafka-on-pulsar 3.0.0.4](https://github.com/streamnative/kop/tree/branch-3.0.0.4) on this install as well.  I only realized when I added some logs to some of the pulsar-broker metrics classes and found the following exception pop up:

```
2024-07-05T16:35:12,221Z [jdk.internal.loader.ClassLoaders$AppClassLoader@5bc2b487] error Uncaught exception in thread pool-15-thread-76: 'org.apache.pulsar.common.util.SimpleTextOutputStream org.apache.pulsar.common.util.SimpleTextOutputStream.write(java.lang.String)'
java.lang.NoSuchMethodError: 'org.apache.pulsar.common.util.SimpleTextOutputStream org.apache.pulsar.common.util.SimpleTextOutputStream.write(java.lang.String)'
		at io.streamnative.pulsar.handlers.kop.stats.PrometheusTextFormatUtil.writeGauge(PrometheusTextFormatUtil.java:32)
		at io.streamnative.pulsar.handlers.kop.stats.PrometheusMetricsProvider.lambda$generate$1(PrometheusMetricsProvider.java:87)
		at java.base/java.util.concurrent.ConcurrentHashMap.forEach(ConcurrentHashMap.java:1603)
		at io.streamnative.pulsar.handlers.kop.stats.PrometheusMetricsProvider.generate(PrometheusMetricsProvider.java:87)
		at org.apache.pulsar.broker.stats.prometheus.PrometheusMetricsGenerator.generateMetrics(PrometheusMetricsGenerator.java:345)
		at org.apache.pulsar.broker.stats.prometheus.PrometheusMetricsGenerator.lambda$renderToBuffer$0(PrometheusMetricsGenerator.java:528)
		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
		at java.base/java.lang.Thread.run(Thread.java:840)
```

I went ahead and removed the kop `.nar` from the installed image and the pulsar-broker metrics began to work again.  Then I pulled down the kop repo and rebuilt it against pulsar 3.0.5.5 and used that new `.nar` and was able to get pulsar-broker metrics still.

I know the kop repo is archived but it looks like that was my issue.

**Comment by merlimat** [comment]
@lhotari PTAL when you have the chance
